### PR TITLE
LuaHook : Pass the character along the effect being applied

### PR DESCRIPTION
**Pull request by challet**
In the initial code the character var was used to add extra targets to the effect before to actually apply it, and so the character was not needed to be passed further.
A LuaHook may have more to do and could need this character var, so it has been added for this purpose.

### CODE REVIEW

File: Barotrauma/BarotraumaShared/SharedSource/StatusEffects/StatusEffect.cs [hunk 1]
```diff
@@ -1371,7 +1371,7 @@ public bool ShouldWaitForInterval(Entity entity, float deltaTime)
             return false;
         }
 
-        public virtual void Apply(ActionType type, float deltaTime, Entity entity, ISerializableEntity target, Vector2? worldPosition = null)
+        public virtual void Apply(ActionType type, float deltaTime, Entity entity, ISerializableEntity target, Vector2? worldPosition = null, character = null)
```
Comment (challet) [inline Barotrauma/BarotraumaShared/SharedSource/StatusEffects/StatusEffect.cs:1374]: Not needed here, but adding it has some consistency with the actual method that will use it. Not a strong opinion though

File: Barotrauma/BarotraumaShared/SharedSource/StatusEffects/StatusEffect.cs [hunk 2]
```diff
@@ -1392,11 +1392,11 @@ public virtual void Apply(ActionType type, float deltaTime, Entity entity, ISeri
             currentTargets.Clear();
             currentTargets.Add(target);
             if (!HasRequiredConditions(currentTargets)) { return; }
-            Apply(deltaTime, entity, currentTargets, worldPosition);
+            Apply(deltaTime, entity, currentTargets, worldPosition, character);
         }
 
         protected readonly List<ISerializableEntity> currentTargets = new List<ISerializableEntity>();
-        public virtual void Apply(ActionType type, float deltaTime, Entity entity, IReadOnlyList<ISerializableEntity> targets, Vector2? worldPosition = null)
+        public virtual void Apply(ActionType type, float deltaTime, Entity entity, IReadOnlyList<ISerializableEntity> targets, Vector2? worldPosition = null, character = null)
```
Comment (challet) [inline Barotrauma/BarotraumaShared/SharedSource/StatusEffects/StatusEffect.cs:1399]: same as the previous one

File: Barotrauma/BarotraumaShared/SharedSource/StatusEffects/StatusEffect.cs [hunk 3]
```diff
@@ -1526,7 +1526,7 @@ protected void Apply(float deltaTime, Entity entity, IReadOnlyList<ISerializable
 
             foreach (string luaHooks in luaHook)
             {
-                var result = GameMain.LuaCs.Hook.Call<bool?>(luaHooks, this, deltaTime, entity, targets, worldPosition);
+                var result = GameMain.LuaCs.Hook.Call<bool?>(luaHooks, this, deltaTime, entity, targets, worldPosition, character);
```
Comment (challet) [inline Barotrauma/BarotraumaShared/SharedSource/StatusEffects/StatusEffect.cs:1529]: last position for retro-compatibility

**Comment by challet** [comment]
By the way, should it be pull-requested here or on the original core repo ?

**Comment by evilfactory** [comment]
Is this required? The LuaHook can just access the character in the targets list.

**Comment by challet** [comment]
I don't know about all the status effects but I'm assuming that for the core ones where the character is involved, yes it will be available in the targets list. Because the effect target would be [one that makes it as so](https://github.com/Regalis11/Barotrauma/blob/60b0e2ae3ed5163c823e8d59bc87a26dac4ab8c4/Barotrauma/BarotraumaShared/SharedSource/Items/Item.cs#L1784-L1808). However, the one I'm working on still need the character without to make it a target (that's however what I ended up doing, to be able to pass it further — I could keep it like that but it feels hacky and inconsistent) : 

```lua
-- Item with the ItemContainer component will fire an OnUse effect when selected (i.e. when their UI is being open)
Hook.Patch("Barotrauma.Items.Components.ItemContainer", "Select", function(instance, ptable))
  if ptable.ReturnValue then
    -- hacky to make the character as the target of the item, but it is needed to pass it further
    -- it also needs the StatusEffect.target to be set on "UseTarget"
    -- until https://github.com/evilfactory/LuaCsForBarotrauma/pull/148 is merged
    instance.item.ApplyStatusEffects(ActionType.OnUse, 0, ptable["character"], nil, ptable["character"])
  end
  return ptable.ReturnValue
end, Hook.HookMethodType.After)
```

It is a general purpose hook thanks to which subsequent actions would be defined through `StatusEffect` and `LuaHook` XML tags.  This hooks with a specific purpose could need to check things involving the character initiating the action.

You may think of something else with better knowledge than mine.

**Comment by evilfactory** [comment]
You are trying to get the character of the calling status effect without the status effect having a target type that targets characters? I assume you can't just add another target because it would interfere with the rest of the status effect? (eg you have some <Affliction/> tag inside or something)
For what I'm thinking, you could instead have two status effect definitions in your XML, each with different targets, so they don't interfere with the another. You are also able to specify multiple targets in a status effect in case that's not the problem. 
Or maybe I'm completely misunderstanding what you are really trying to do.

**Comment by challet** [comment]
The problem you tell would indeed be one that might happen. But I was thinking more in an abstracted way that the initiator of the action cannot be the target (in that specific example of opening a container).

What I'm trying to actually achieve is to add some status effect triggers that would allow other mods to build upon it. But maybe I'm overthinking it, and they would just need to use the `Character` target.

**Comment by challet** [comment]
Putting this on hold for now

**Pull request closed without merging**